test/remote/gateways/remote_redsys_sha256_test.rb

```diff
@@ -18,6 +18,27 @@ def test_successful_purchase
     assert_equal 'Transaction Approved', response.message
   end
 
+  def test_successful_purchase_threeds2
+    xid = '97267598-FAE6-48F2-8083-C23433990FBC'
+    ds_transaction_id = '97267598-FAE6-48F2-8083-C23433990FBC'
+    version = '2.1.0'
+
+    response = @gateway.purchase(
+      100,
+      @credit_card,
+      @options.merge(
+        three_d_secure: {
+          version: version,
+          ds_transaction_id: ds_transaction_id,
+          xid: xid
+        }
+      )
+    )
+
+    assert_success response
+    assert_equal 'Transaction Approved', response.message
+  end
+
```

Comment:
Given that we've had various issues with strict ordering of XML requests, I'd suggest testing this with the fields that could be added after the mpi element. Were you able to find a schema?

Reply:
yes, by adding ?WSDL to the URL it shows the schema, they're all marked "literal", strict ordering/type check is not enforced at the structural level. I'll add some smoke tests for our peace of mind

Reply:
Nice!